Resource Group FlyWay Migration does not work with Trino Secrets (Server doesnt start)

Hello,

We currently use (secrets)[https://trino.io/docs/current/security/secrets.html] to manage our secrets within the Trino Environment. This has been working fine until our recent upgrade to 0.365. We noticed that there is an issue with the flyway migration. We know that a (change)[https://github.com/trinodb/trino/commit/cbd1027327401c708222e4378210f80d306155c3] was done for 0.364. We tried out the 0.365 version using secrets and found out the Trino Server is unable to start. The follow is the errors that we get.

```
021-12-13T18:40:43.617Z	INFO	main	io.trino.plugin.resourcegroups.db.FlywayMigration	Performing migrations...
2021-12-13T18:40:43.703Z	ERROR	main	io.trino.server.Server	Unable to obtain connection from database (jdbc:mysql://${ENV:CREDENTIALS_ENDPOINT}/${ENV:RESOURCE_GROUP_DB_NAME}?user=${ENV:CREDENTIALS_PRESTOPROXY_DB_USERNAME}&password=******************************************&useSSL=false) for user 'null': Cannot load connection class because of underlying exception: com.mysql.cj.exceptions.WrongArgumentException: Failed to parse the host:port pair '${ENV:CREDENTIALS_ENDPOINT}'.
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
SQL State  : 08001
Error Code : 0
Message    : Cannot load connection class because of underlying exception: com.mysql.cj.exceptions.WrongArgumentException: Failed to parse the host:port pair '${ENV:CREDENTIALS_ENDPOINT}'.

org.flywaydb.core.internal.exception.FlywaySqlException: Unable to obtain connection from database (jdbc:mysql://${ENV:CREDENTIALS_ENDPOINT}/${ENV:RESOURCE_GROUP_DB_NAME}?user=${ENV:CREDENTIALS_PRESTOPROXY_DB_USERNAME}&password=******************************************&useSSL=false) for user 'null': Cannot load connection class because of underlying exception: com.mysql.cj.exceptions.WrongArgumentException: Failed to parse the host:port pair '${ENV:CREDENTIALS_ENDPOINT}'.
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
SQL State  : 08001
Error Code : 0
Message    : Cannot load connection class because of underlying exception: com.mysql.cj.exceptions.WrongArgumentException: Failed to parse the host:port pair '${ENV:CREDENTIALS_ENDPOINT}'.

	at org.flywaydb.core.internal.jdbc.JdbcUtils.openConnection(JdbcUtils.java:67)
	at org.flywaydb.core.internal.jdbc.JdbcConnectionFactory.<init>(JdbcConnectionFactory.java:70)
	at org.flywaydb.core.Flyway.execute(Flyway.java:504)
	at org.flywaydb.core.Flyway.migrate(Flyway.java:170)
	at io.trino.plugin.resourcegroups.db.FlywayMigration.migrate(FlywayMigration.java:38)
	at io.trino.plugin.resourcegroups.db.DbResourceGroupConfigurationManagerFactory.create(DbResourceGroupConfigurationManagerFactory.java:42)
	at io.trino.execution.resourcegroups.InternalResourceGroupManager.setConfigurationManager(InternalResourceGroupManager.java:163)
	at io.trino.execution.resourcegroups.InternalResourceGroupManager.loadConfigurationManager(InternalResourceGroupManager.java:147)
	at io.trino.server.Server.doStart(Server.java:131)
	at io.trino.server.Server.lambda$start$0(Server.java:78)
	at io.trino.$gen.Trino_365_1____20211213_184022_1.run(Unknown Source)
	at io.trino.server.Server.start(Server.java:78)
	at io.trino.server.TrinoServer.main(TrinoServer.java:38)
Caused by: java.sql.SQLNonTransientConnectionException: Cannot load connection class because of underlying exception: com.mysql.cj.exceptions.WrongArgumentException: Failed to parse the host:port pair '${ENV:CREDENTIALS_ENDPOINT}'.
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:110)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:97)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:89)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:63)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:73)
	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:79)
	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:131)
	at com.mysql.cj.jdbc.NonRegisteringDriver.connect(NonRegisteringDriver.java:221)
	at org.flywaydb.core.internal.jdbc.DriverDataSource.getConnectionFromDriver(DriverDataSource.java:268)
	at org.flywaydb.core.internal.jdbc.DriverDataSource.getConnection(DriverDataSource.java:232)
	at org.flywaydb.core.internal.jdbc.JdbcUtils.openConnection(JdbcUtils.java:52)
	... 12 more
Caused by: com.mysql.cj.exceptions.UnableToConnectException: Cannot load connection class because of underlying exception: com.mysql.cj.exceptions.WrongArgumentException: Failed to parse the host:port pair '${ENV:CREDENTIALS_ENDPOINT}'.
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at com.mysql.cj.exceptions.ExceptionFactory.createException(ExceptionFactory.java:61)
	at com.mysql.cj.exceptions.ExceptionFactory.createException(ExceptionFactory.java:105)
	... 16 more
Caused by: com.mysql.cj.exceptions.WrongArgumentException: Failed to parse the host:port pair '${ENV:CREDENTIALS_ENDPOINT}'.
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at com.mysql.cj.exceptions.ExceptionFactory.createException(ExceptionFactory.java:61)
	at com.mysql.cj.exceptions.ExceptionFactory.createException(ExceptionFactory.java:105)
	at com.mysql.cj.conf.ConnectionUrlParser.parseHostPortPair(ConnectionUrlParser.java:502)
	at com.mysql.cj.conf.ConnectionUrlParser.buildHostInfoResortingToGenericSyntaxParser(ConnectionUrlParser.java:440)
	at com.mysql.cj.conf.ConnectionUrlParser.parseAuthoritySegment(ConnectionUrlParser.java:269)
	at com.mysql.cj.conf.ConnectionUrlParser.parseAuthoritySection(ConnectionUrlParser.java:183)
	at com.mysql.cj.conf.ConnectionUrlParser.getHosts(ConnectionUrlParser.java:653)
	at com.mysql.cj.conf.ConnectionUrl$Type.getConnectionUrlInstance(ConnectionUrl.java:194)
	at com.mysql.cj.conf.ConnectionUrl.getConnectionUrlInstance(ConnectionUrl.java:280)
	at com.mysql.cj.jdbc.NonRegisteringDriver.connect(NonRegisteringDriver.java:195)
	... 15 more
Caused by: java.lang.NumberFormatException: For input string: "CREDENTIALS_ENDPOINT}"
	at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.base/java.lang.Integer.parseInt(Integer.java:652)
	at java.base/java.lang.Integer.parseInt(Integer.java:770)
	at com.mysql.cj.conf.ConnectionUrlParser.parseHostPortPair(ConnectionUrlParser.java:500)
	... 22 more
```

@posulliv 

An update here. Thank you for the patch. Unfortnuately, I ran into issues but I was able to work around it.

We used to have this config only in our resource_groups.properties

```
resource-groups.config-db-url=jdbc:mysql://\${ENV:CREDENTIALS_ENDPOINT}/\${ENV:RESOURCE_GROUP_DB_NAME}?user=\${ENV:CREDENTIALS_PRESTOPROXY_DB_USERNAME}&password=\${ENV:CREDENTIALS_PRESTOPROXY_DB_PASSWORD}&useSSL=false
```

I then applied your fix ontop of 365 and got the following

```
1) [Guice/ErrorInCustomProvider]: NullPointerException: null username
  at DbResourceGroupsModule.create(DbResourceGroupsModule.java:44)
  at DaoProvider.<init>(DaoProvider.java:31)
      \_ for 2nd parameter jdbi
  while locating DaoProvider
  at DbResourceGroupsModule.configure(DbResourceGroupsModule.java:34)
  while locating ResourceGroupsDao

Learn more:
  https://github.com/google/guice/wiki/ERROR_IN_CUSTOM_PROVIDER
Caused by: NullPointerException: null username
        at java.base/Objects.requireNonNull(Objects.java:246)
        at Jdbi.create(Jdbi.java:138)
        at DbResourceGroupsModule.create(DbResourceGroupsModule.java:44)
        at DbResourceGroupsModule$$FastClassByGuice$$864557.GUICE$TRAMPOLINE(<generated>)
        at DbResourceGroupsModule$$FastClassByGuice$$864557.apply(<generated>)
        at ProviderMethod$FastClassProviderMethod.doProvision(ProviderMethod.java:260)
        at ProviderMethod.doProvision(ProviderMethod.java:171)
        at InternalProviderInstanceBindingImpl$CyclicFactory.provision(InternalProviderInstanceBindingImpl.java:185)
        at InternalProviderInstanceBindingImpl$CyclicFactory.access$300(InternalProviderInstanceBindingImpl.java:139)
        at InternalProviderInstanceBindingImpl$CyclicFactory$1.call(InternalProviderInstanceBindingImpl.java:169)
        at ProvisionListenerStackCallback$Provision.provision(ProvisionListenerStackCallback.java:109)
        at LifeCycleModule.provision(LifeCycleModule.java:54)
        at ProvisionListenerStackCallback$Provision.provision(ProvisionListenerStackCallback.java:117)
        at ProvisionListenerStackCallback.provision(ProvisionListenerStackCallback.java:66)
        at InternalProviderInstanceBindingImpl$CyclicFactory.get(InternalProviderInstanceBindingImpl.java:164)
        at ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40)
        at SingletonScope$1.get(SingletonScope.java:169)
        at InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:45)
        at SingleParameterInjector.inject(SingleParameterInjector.java:40)
        at SingleParameterInjector.getAll(SingleParameterInjector.java:60)
        at ConstructorInjector.provision(ConstructorInjector.java:113)
        at ConstructorInjector.access$000(ConstructorInjector.java:33)
        at ConstructorInjector$1.call(ConstructorInjector.java:98)
        at ProvisionListenerStackCallback$Provision.provision(ProvisionListenerStackCallback.java:109)
        at LifeCycleModule.provision(LifeCycleModule.java:54)
        at ProvisionListenerStackCallback$Provision.provision(ProvisionListenerStackCallback.java:117)
        at ProvisionListenerStackCallback.provision(ProvisionListenerStackCallback.java:66)
        at ConstructorInjector.construct(ConstructorInjector.java:93)
        at ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:296)
        at BoundProviderFactory.get(BoundProviderFactory.java:58)
        at ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40)
        at SingletonScope$1.get(SingletonScope.java:169)
        at InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:45)
        at InternalInjectorCreator.loadEagerSingletons(InternalInjectorCreator.java:213)
        at InternalInjectorCreator.injectDynamically(InternalInjectorCreator.java:186)
        at InternalInjectorCreator.build(InternalInjectorCreator.java:113)
        at Guice.createInjector(Guice.java:87)
        at Bootstrap.initialize(Bootstrap.java:275)
        at DbResourceGroupConfigurationManagerFactory.create(DbResourceGroupConfigurationManagerFactory.java:64)
        at InternalResourceGroupManager.setConfigurationManager(InternalResourceGroupManager.java:163)
```

Seems like we ran into an issue here.

https://github.com/trinodb/trino/blob/7b0e9b2bd925700456293587b274c5a219ce56dd/plugin/trino-resource-group-managers/src/main/java/io/trino/plugin/resourcegroups/db/DbResourceGroupsModule.java#L44

Now since we did not define a user and password, this crashed. This was working in the previous releases with only the URL encoded.

I got around this now by adding the following to the config

```
    resource-groups.config-db-user=\${ENV:CREDENTIALS_PRESTOPROXY_DB_USERNAME}

    resource-groups.config-db-password=\${ENV:CREDENTIALS_PRESTOPROXY_DB_PASSWORD}
``` 

I think this should still be fixed going forward as it looks like a regression even with the patch.

--- a/plugin/trino-resource-group-managers/src/main/java/io/trino/plugin/resourcegroups/db/DbResourceGroupConfigurationManagerFactory.java
+++ b/plugin/trino-resource-group-managers/src/main/java/io/trino/plugin/resourcegroups/db/DbResourceGroupConfigurationManagerFactory.java
@@ -27,6 +27,8 @@ import org.weakref.jmx.guice.MBeanModule;
 
 import java.util.Map;
 
+import static io.airlift.configuration.ConfigurationUtils.replaceEnvironmentVariables;
+
 public class DbResourceGroupConfigurationManagerFactory
         implements ResourceGroupConfigurationManagerFactory
 {
@@ -39,6 +41,7 @@ public class DbResourceGroupConfigurationManagerFactory
     @Override
     public ResourceGroupConfigurationManager<?> create(Map<String, String> config, ResourceGroupConfigurationManagerContext context)
     {
+        config = replaceEnvironmentVariables(config);
         FlywayMigration.migrate(new ConfigurationFactory(config).build(DbResourceGroupConfig.class));
         Bootstrap app = new Bootstrap(
                 new MBeanModule(),

